src/buffer/out/textBuffer.cpp

```diff
@@ -166,6 +166,19 @@ ROW& TextBuffer::_getRowByOffsetDirect(size_t offset)
     return *reinterpret_cast<ROW*>(row);
 }
 
+// Returns the "user-visible" index of the last committed row, which can be used
+// to short-circuit some algorithms that try to scan the entire buffer.
+// Returns 0 if no rows are committed in.
+til::CoordType TextBuffer::_estimateOffsetOfLastCommittedRow() const noexcept
+{
+    const auto lastRowOffset = (_commitWatermark - _buffer.get()) / _bufferRowStride;
+    // This subtracts 2 from the offset to account for the:
+    // * scratchpad row at offset 0, whereas regular rows start at offset 1.
+    // * fact that _commitWatermark points _past_ the last committed row,
+    //   but we want to return an index pointing at the last row.
+    return std::max(0, gsl::narrow_cast<til::CoordType>(lastRowOffset - 2));

```

Comment:
I realized that the
```cpp
const auto y = (_commitWatermark - _buffer.get() - 1) / _bufferRowStride;
```
calculation is effectively the same as
```cpp
const auto x = (_commitWatermark - _buffer.get()) / _bufferRowStride;
const auto y = x - 1;
```
and so it kind of merges with the other -1 we already had to form -2.
I also felt like it's neat because now it's only 2 lines of code and the `std::max` ensures it's impossible to return negative indices.

Reply:
Hmmmmm I agree with you, but I probably should have been the one to commit the suggestion ;)

Reply:
Ah sorry, I thought this would take some load off your shoulders.